Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/checkins/checkout/ContactJournalEntryCreator.kt

```diff
@@ -0,0 +1,76 @@
+package de.rki.coronawarnapp.presencetracing.checkins.checkout
+
+import dagger.Reusable
+import de.rki.coronawarnapp.contactdiary.model.ContactDiaryLocation
+import de.rki.coronawarnapp.contactdiary.model.ContactDiaryLocationVisit
+import de.rki.coronawarnapp.contactdiary.model.DefaultContactDiaryLocation
+import de.rki.coronawarnapp.contactdiary.model.DefaultContactDiaryLocationVisit
+import de.rki.coronawarnapp.contactdiary.storage.repo.ContactDiaryRepository
+import de.rki.coronawarnapp.eventregistration.checkins.CheckIn
+import de.rki.coronawarnapp.eventregistration.checkins.split.splitByMidnightUTC
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toLocalDateUtc
+import kotlinx.coroutines.flow.first
+import org.joda.time.Duration
+import org.joda.time.Seconds
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.math.roundToLong
+
+@Reusable
+class ContactJournalEntryCreator @Inject constructor(
+    private val diaryRepository: ContactDiaryRepository
+) {
+
+    suspend fun createEntry(checkIn: CheckIn) {
+        if (checkIn.createJournalEntry) {
+            Timber.d("Creating journal entry for %s", this)
+
+            // 1. Create location if missing
+            val traceLocationIdString = checkIn.traceLocationId.utf8()
+            val location: ContactDiaryLocation = diaryRepository.locations.first()
+                .find { it.traceLocationID == traceLocationIdString } ?: checkIn.toLocation(traceLocationIdString)
+
+            // 2. Split CheckIn by Midnight UTC
+            val splitCheckIns = checkIn.splitByMidnightUTC()
+            Timber.d("Split %s into %s ", this, splitCheckIns)
+
+            // 3. Create LocationVisit
+            splitCheckIns
+                .map { it.toLocationVisit(location) }
+                .forEach { diaryRepository.addLocationVisit(it) }
```

Comment:
Something with the ID check and visit generation is still off.

When checking into the same location twice.
* Scan QR, checkout after a few seconds
* Scan QR, checkout after a few seconds.
I endet up with two location entries, on overview, a single location on the day and no visit.
When I try to create a visit for that location, it doesn't expand the entry, but generates multiple visits on the overview.

![Screenshot from 2021-04-08 17-47-11](https://user-images.githubusercontent.com/1439229/114057602-fcde3400-9892-11eb-8de5-3ab2bb24668d.png)
![Screenshot from 2021-04-08 17-46-53](https://user-images.githubusercontent.com/1439229/114057618-ff408e00-9892-11eb-860d-8590d1b59475.png)
![Screenshot from 2021-04-08 17-46-44](https://user-images.githubusercontent.com/1439229/114057619-ffd92480-9892-11eb-8ad4-bfe669771fa3.png)
![Screenshot from 2021-04-08 17-46-36](https://user-images.githubusercontent.com/1439229/114057622-ffd92480-9892-11eb-97eb-9d22ee8809b4.png)
![Screenshot from 2021-04-08 17-52-35](https://user-images.githubusercontent.com/1439229/114057847-3b73ee80-9893-11eb-8c24-842b4b1fc94b.png)

Reply:
check this for details: https://github.com/corona-warn-app/cwa-app-tech-spec/pull/30/commits/693e85f79a918bef112d63693fd36a16b7ec8d99

adjusted.